docs: classic worker (#7203)
vitejs/vite: docs/guide/features.md

## Web Workers

### Import with Constructors

> A web worker script can also be imported using [`new Worker()`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker) and [`new SharedWorker()`](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker/SharedWorker). Compared to the worker suffixes, this syntax leans closer to the standards and is the **recommended** way to create workers.

Reviewer: ```suggestion
A web worker script can be imported using [`new Worker()`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker) and [`new SharedWorker()`](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker/SharedWorker). Compared to the worker suffixes, this syntax leans closer to the standards and is the **recommended** way to create workers.
```

## Web Workers

> ### Import with query suffixes

Reviewer: I just remembered that the Vite/Vue docs use title case for this
```suggestion
### Import with Query Suffixes
```

> ### imported by appending `?worker` or `?sharedworker`

Reviewer: I think we can simplify this part as
```suggestion
### `?worker` and `?sharedworker`
```

import MyWorker from './worker?worker&inline'
```

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.

Author: > `?worker` would be deprecated soon 

If so, I think we still have something to do. For example

```js
import url from "./worker?url"

new Worker(url)
```

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.

Author: or get give a issues for dynamic express in worker url too.

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.

Author: IIUC. `import.meta.*` only run in `<script type="module">`?

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.

Reviewer: > IIUC. `import.meta.*` only run in `<script type="module">`?

Yeah.

Re `?url`, I wonder if `new URL('./worker.js', import.meta.url)` works today? Or maybe it needs to be in a `new Worker` for it to make sense. I think you know more about these than I do 😄

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.

Author: hhh, I think this is good point. This syntax is closer to native. 

I really can't think of another paragraph with a good title 😅

Web worker is `?worker`
Or divide them into two three-level headings

> ### imported by appending `?worker` or `?sharedworker`

> ### new Worker()

Detect the case where `{ type: 'classic' }` is not provided or `{ type: 'classic' }` is passed with a static string and use classic workers.

```ts
const worker = new Worker(new URL('./worker.js', import.meta.url), {
  type: 'classic'
})
const worker = new SharedWorker(new URL('./worker.js', import.meta.url), {
  type: 'classic'
})
```

::: tip Note

If a dynamic expression appears, vite will issue an error. And give `/* @vite-ignore */` out of Vite processing. Because use dynamic expression, vite will copy the code of `/@vite/env` to the worker code head.

```ts
const generateWorkerOption = () => ({ type: 'classic' })
// use dy dynamic expression will got a vite process error.
const worker = new Worker(
  new Url('./worker.js', import.meta.url),
  generateWorkerOption()
)

> ```

Reviewer: I think we can simplify this part with "the worker options must be static for Vite to analyze it. if it contains dynamic expressions, `/* @vite-ignore */` can be used to opt-out of Vite processing". 

I don't think we need to explain `/@vite/env` as that's an implementation detail. We can also simplify the code block with `{ type: workerType }` (uses variables) perhaps and show how to use the `/* @vite-ignore */` comment. I think `Url` should be `URL` too.

import MyWorker from './worker?worker&inline'
```

> ### Import with constructors

Reviewer: ```suggestion
### Import with Constructors
```

> ### Import with Constructors

Reviewer 2: Shall we put this section before `Import with Query Suffixes` since it said this is the recommended way to create workers so readers would feel the docs emphasizes this than suffixes?

Reviewer: I think this makes sense, but I didn't suggest it earlier because there's no way to do `?inline` with constructors yet (if i'm not mistaken). But I'm fine either way too.

Reviewer: I think the end goal is to remove query suffixes everywhere for Vite, but we can't put `inline` in worker options anyways since that's non-standard. I'm actually not sure what's the right call here but it seems like it needs more discussion.

Author: yep, this is compile-time option don't let it into runtime option. 😊 so how about 
```ts
const worker = new Worker(new URL(/* #__INLINE__ */'./worker.js', import.meta.url))
```

Reviewer: I think we would want to refrain from special comments too 😬 IIRC Rollup has a history of avoiding that too (e.g. `webpackChunkName`) and only supported pure comments thus far. I think Vite should follow suit too and find some alternatives. I'm not sure what yet, but maybe we can create a new issue for it to toss some ideas, and have it discussed soon.

Author: ok, https://github.com/vitejs/vite/issues/7352 let us discuss this.

A web worker script can also be imported using [`new Worker()`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker) and [`new SharedWorker()`](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker/SharedWorker). Compared to the worker suffixes, this syntax leans closer to the standards and is the recommended way to create workers.

```ts
const worker = new Worker(new URL('./worker.js', import.meta.url))
```

> The worker constructor also accepts options, which can be used to create "classic" workers:
> 
> ```ts
> const worker = new Worker(new URL('./worker.js', import.meta.url), {
>   type: 'classic'
> })
> ```

Reviewer: IIRC Vite treats workers without options as classic by default? Should we show an example for module workers instead?

Note: GitHub formats the below weirdly.

```suggestion
The worker constructor also accepts options, which can be used to create "module" workers:

```ts
const worker = new Worker(new URL('./worker.js', import.meta.url), {
  type: 'module'
})
```
```

import MyWorker from './worker?worker&inline'
```

> ### new Worker()

Reviewer: ```suggestion
### `new Worker()` and `new SharedWorker()`
```

Author: It looks more concise. I also thought of a name. I don't know whether it's good or not. :p

`es suffixes` / `native`

Author: Because `the worker suffixes` is used for short below `?worker`. I think it might be better to use abbreviations in the title as soon as possible?

Reviewer: Yeah, I've thought about that too and am also not sure what's the right name for it. Maybe `Import with suffixes` / `Import with constructors`

Author: how about `import with es query`

Reviewer: I don't think we need to mention "es", but query would be a good alternative too. Maybe "Import with query suffixes"? But I'm also fine with just "suffixes".

Author: great word

> A web worker script can be imported by `new Worker()`. The difference between this syntax and imported by appending `?worker` or `?sharedworker` is that this one can use worker options. So we can create a `classic worker` by this syntax.
> 
> Detect the case where `{ type: 'classic' }` is not provided or `{ type: 'classic' }` is passed with a static string and use classic workers.

Reviewer: ```suggestion
A web worker script can also be imported using [`new Worker()`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker) and [`new SharedWorker()`](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker/SharedWorker). Compared to the worker suffixes, this syntax leans closer to the standards and is the recommended way to create workers.
```

Summarized this part a little as the "difference" part was a bit verbose 😬

Author: Don't need to remind `worker options` that they need static strings?

> ```ts
> const worker = new Worker(new URL('./worker.js', import.meta.url), {
>   type: 'classic'
> })
> const worker = new SharedWorker(new URL('./worker.js', import.meta.url), {
>   type: 'classic'
> })
> ```

Reviewer: We can simplify this part and explain `type: classic` later. Note: Github is not liking the formatting below.
```suggestion
```ts
const worker = new Worker(new URL('./worker.js', import.meta.url))
```\
```

> 

Reviewer: Over here we can touch on worker options:
```suggestion
The worker constructor also accepts options, which can be used to create "classic" workers:

```ts
const worker = new Worker(new URL('./worker.js', import.meta.url), {
  type: 'classic'
})
```\
```

> ::: tip Note
> 
> the worker options must be static for Vite to analyze it. if it contains dynamic expressions, `/* @vite-ignore */` can be used to opt-out of Vite processing
> 
> ```ts
> const generateWorkerOption = () => ({ type: 'classic' })
> // use dy dynamic expression will got a vite process error.
> const worker = new Worker(
>   new URL('./worker.js', import.meta.url),
>   generateWorkerOption()
> )
> ```
> 
> No matter which type of worker is used, the same [worker configuration](/config/#worker-options) will be used for bundle.
> 
> :::

Reviewer: I think we can skip the `vite-ignore` part here, I wonder what others think about this. Because it's an edge case and if Vite hits something like this, the dev server would already suggest using `@vite-ignore`.

If we do remove that, we can simplify this entire part to explain about the worker configuration.

```suggestion
See [Worker Options](/config/#worker-options) for details on configuring the bundling of all workers.
```

Author: good point.
